beps/0006-scaffolder-action-rollback/README.md

```diff
@@ -1,84 +1,303 @@
+2. Attempt to rollback "Setup branch protection rules" by removing the rules
+3. Attempt to rollback "Create repository" by deleting the repository
+
+Each checkpoint's rollback function handles its specific resource cleanup, ensuring atomic operations with clear boundaries.
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant Backstage UI
+    participant NunjucksWorkflowRunner
+    participant Action as "GitHub Action"
+    participant CP1 as "Checkpoint 1<br/>(Create Repo)"
+    participant CP2 as "Checkpoint 2<br/>(Branch Rules)"
+    participant CP3 as "Checkpoint 3<br/>(Team Access)"
+    participant CP4 as "Checkpoint 4<br/>(Create PR)"
+
+    User->>Backstage UI: Start Template Execution
+    Backstage UI->>NunjucksWorkflowRunner: Execute Template
+
+    Note over NunjucksWorkflowRunner: Normal Execution Flow
 
-- when user manually decide to perform this action.
-- task has to be recovered and TaskRecoverStrategy set to 'rollback'
+    NunjucksWorkflowRunner->>Action: Execute GitHub Action
+    Action->>CP1: Process (success)
+    CP1-->>Action: Complete with data
+    Action->>CP2: Process (success)
+    CP2-->>Action: Complete with data
+    Action->>CP3: Process (success)
+    CP3-->>Action: Complete with data
+    Action->>CP4: Process (fails)
+    CP4--xAction: Error
+    Action--xNunjucksWorkflowRunner: Action Failed
+
+    Note over NunjucksWorkflowRunner: Rollback Flow (LIFO order by checkpoint)
+
+    NunjucksWorkflowRunner->>CP3: Rollback (Team Access)
+    CP3-->>NunjucksWorkflowRunner: Rollback Complete
+    NunjucksWorkflowRunner->>CP2: Rollback (Branch Rules)
+    CP2-->>NunjucksWorkflowRunner: Rollback Complete
+    NunjucksWorkflowRunner->>CP1: Rollback (Repository)
+    CP1-->>NunjucksWorkflowRunner: Rollback Complete
+
+    NunjucksWorkflowRunner-->>Backstage UI: Template Execution Failed
+    Backstage UI-->>User: Display Error & Rollback Status
+```
 
 ## Design Details
 
-We are going to introduce an extra function in action API which will look like:
+### Extending Checkpoint API for Rollback
+
+The checkpoint API will be extended to include an optional rollback function that can utilize the data saved by the checkpoint:
+
+```typescript
+/**
+ * Extended Checkpoint API with rollback support
+ */
+async checkpoint<T>(
+  key: string,
+  fn: () => Promise<T>,
+  rollback?: (data: T) => Promise<void>
+): Promise<T>;
+```
+
+This allows the checkpoint to save data during forward execution that can be used to properly rollback the operation if needed.
+
+Example usage:
+
+```typescript
+// Repository creation with rollback
+await ctx.checkpoint(
+  'repository-created',
+  async () => {
+    const { owner, repo } = ctx.input;
+    const repository = await githubClient.repos.create({ owner, repo });
+    return {
+      repositoryId: repository.id,
+      owner,
+      repo,
+    };
+  },
+  async data => {
+    // Rollback function using saved checkpoint data
+    const { owner, repo } = data;
```

Comment:
I wonder if it's worth calling out this with an example of using the `data` here, I was a little confused that this was like the initial input to the action, and not the response of the `fn`.

Maybe just add like `ctx.logger('Removing repository, { id: data.repository.id }')` or something?